SOF-2298: Synchronize-only ingest flow

KvelaGorrrrnio opened the pull request:
_**Note:** For the web client to behave correctly, the PR https://github.com/tv2/sofie-web-client/pull/272 is needed._

The PR introduces a new approach for handling ingest events observed through MongoDB that Sofie Core generates.

The approach used is to analyze the entire state of a rundown, when a sequence of observed ingest events for the rundown (or more rundowns) ends. By doing it this way, the code that runs at startup to ensure correct state and the code that runs when ingest events are received is the same.

The code analyzes the ingest data for the rundown against the existing rundown and classifies any changes to the rundown metadata, along with create, update and delete for segments and parts.

If a segment is affected (created, updated or deleted), then part changes under that segment is ignored, since the segment includes the new state of the parts. So the part changes are redundant and will most likely throw errors, when applied to the rundown - e.g. adding a part that is already there from the segment update.

One of the trickiest parts regarding handling ingest events from Sofie Core is to properly manage changes for on air segments and parts. The way changes to an on air segment is handled is by allowing all changes, since they are metadata changes.
Deleting an on air part results in an unsynced on air part, and an update results in an unsynced on air part, followed by a synced version of the part that will receive future updates.

To determine whether a part is updated or not, `EntityChangeDetector` is used, where all relevant attributes are compared. Complex objects are stringified and key-value pairs with the key `id` is removed. The removal of `id` attributes is because random ids were generated for sisyfos layers for each ingest, so a fixed amount of parts would always be detected as changed.

KvelaGorrrrnio committed:
chore: Adds service for generating, persisting and emitting actions.

KvelaGorrrrnio committed:
test: Adds ingested entities to EntityTestFactory.

KvelaGorrrrnio committed:
chore: Adds EntityChangeDetector for detecting changes between entities and ingested entities.

KvelaGorrrrnio committed:
chore: Adds IngestRundownSynchronizer for analyzing changes between a rundown and an ingested rundown.

KvelaGorrrrnio committed:
fix: Replaces priority queue ingest flow with a syncrhonize-only strategy.

Technically, IngestDataChangedService is replaced by IngestDataChangeService.

KvelaGorrrrnio committed:
chore: Makes resulting array of getEntities methods for Rundown, Segment and Part read-only.

KvelaGorrrrnio committed:
fix: Repository cache is updated after database changes are applied.

KvelaGorrrrnio committed:
chore: Adds more information to part already exists exception message when trying to add part to segment.

KvelaGorrrrnio committed:
fix: Includes unplanned pieces when updating from ingested entities.

KvelaGorrrrnio committed:
fix: Updates active cursor when the on air part is updated.

KvelaGorrrrnio committed:
test: Update data for given to EntityTestFactory method.

KvelaGorrrrnio committed:
fix: Keep unplanned parts when ingesting.

LindvedKrvang commented on src/model/entities/test/entity-test-factory.ts:
```diff
@@ -110,7 +142,7 @@ export class EntityTestFactory {
       start: 0,
       pieceLifespan: PieceLifespan.WITHIN_PART,
-      isPlanned: false,
+      isPlanned: true,
```
Remember to dobbeltcheck that this doesn't break any existing tests

KvelaGorrrrnio replied:
What should I test besides running the tests? :)

LindvedKrvang replied:
![image](https://github.com/user-attachments/assets/93483c20-70ca-442e-a14f-22271ad3bca3)

KvelaGorrrrnio committed:
fix: Unsync parts when content is updated (e.g. changd source of split screen).

KvelaGorrrrnio committed:
fix: On air parts that are unsynced by updates creates a new part with the updated data.

KvelaGorrrrnio committed:
fix: Use non-updated part when unsyncing on air part from part update.

KvelaGorrrrnio committed:
refactor: Split IngestRundownSynchronizer.synchronizeRundown up into smaller methods.

KvelaGorrrrnio committed:
test: Moves describe under the method under test-describe.

KvelaGorrrrnio committed:
chore: Trying to update an on air part with the IngestedEntityToEntityMapper should return the on air part unchanged.

KvelaGorrrrnio committed:
fix: Changed order of ingest change type application to the rundown and keeps created and deleted events for on air part even though the segment is affected.

KvelaGorrrrnio committed:
refactor: Changed event emitting guard to a filter.

KvelaGorrrrnio committed:
fix: Overwritten planned pieces on off air parts are kept when processing ingest changes.

Prior to this, the overwritten pieces were lost when the part was reset.

KvelaGorrrrnio committed:
fix: Keep unplanned parts on next part when applying ingest changes.

KvelaGorrrrnio committed:
fix: Change detection for pieces now operatoes on ingested pieces from the part.

The issue with comparing the pieces of the part and the ingested pieces of the ingested part was that a part can have overwritten pieces, which would would mark it as updated even though no ingested changes should be applied.

KvelaGorrrrnio committed:
test: Corrected describe message.

RasmusAlbrektsen commented on src/business-logic/services/ingest-data-change-service.ts:
```diff
@@ -0,0 +1,297 @@
+    return { deletedPartsInfo, deletedSegmentsInfo }
+  }
+
+  private emitEventsFromRundownSynchronizeResult(rundown: Rundown, rundownSynchronizeResult: RundownSynchronizeResult, deletedSegmentsInfo: readonly DeletedSegmentInfo[], deletedPartsInfo: readonly DeletedPartInfo[]): void {
```
These parameter are not nice to read :-D 
Maybe we should consider a different way of showing them and add it to linting?

RasmusAlbrektsen replied:
Generally the line length we allow is a bit much.

LindvedKrvang replied:
The last two parameters can be combine in the type `DeletedInfo`.

KvelaGorrrrnio replied:
I have used `DeletedInfo` instead.

RasmusAlbrektsen commented on src/model/entities/part.ts:
```diff
@@ -204,30 +204,28 @@ export class Part {
-      }
-      this.replacedPlannedPieces.push(piece)
-    })
+  private doesPiecesOverlap(pieceA: Piece, pieceB: Piece): boolean {
```
Personal opinion: 
I'm not a fan of how `does` is making it in when it isn't grammatically correct, in this case it would be `do`. I understand that we are used to doing certain things when naming boolean return functions.

KvelaGorrrrnio replied:
I agree. I have updated it to `doPiecesOverlap`.

LindvedKrvang commented on src/business-logic/services/ingest-data-change-service.ts:
```diff
@@ -221,34 +233,51 @@ export class IngestDataChangeService implements DataChangeService {
-    rundownSynchronizeResult.updatedParts.forEach(part => rundown.updatePart(part))
-    rundownSynchronizeResult.deletedParts.forEach(part => rundown.removePartFromSegment(part.id))
-    return rundownSynchronizeResult.deletedSegments.map(segment => {
+  private applyRundownSynchronizeResult(rundown: Rundown, rundownSynchronizeResult: RundownSynchronizeResult): { deletedPartsInfo: DeletedPartInfo[], deletedSegmentsInfo: DeletedSegmentInfo[] } {
```
You have a nice `DeletedInfo` type that you can use here :)

KvelaGorrrrnio replied:
Updated :)

LindvedKrvang commented on src/business-logic/services/ingest-rundown-synchronizer.ts:
```diff
@@ -23,23 +24,52 @@ export class IngestRundownSynchronizer {
-        return updatedSegments
-      }
-      const segment: Segment | undefined = rundown.getSegments().find(segment => segment.id === ingestedSegment.id)
+  public getUpdatedSegments(segments: readonly Segment[], ingestedSegments: readonly IngestedSegment[]): readonly Segment[] {
```
Should be private

KvelaGorrrrnio replied:
Updated :)

LindvedKrvang commented on src/business-logic/services/ingest-rundown-synchronizer.ts:
```diff
@@ -51,48 +81,65 @@ export class IngestRundownSynchronizer {
-
-    const deletedSegmentIds: ReadonlySet<string> = new Set(deletedSegments.map(segment => segment.id))
-    const deletedParts: readonly Part[] = parts.filter(part => part.isPlanned && !part.isUnsynced() && !ingestedPartIds.has(part.id) && !deletedSegmentIds.has(part.getSegmentId()))
+  public getDeletedSegments(segments: readonly Segment[], ingestedSegments: readonly IngestedSegment[]): readonly Segment[] {
```
Should be private

KvelaGorrrrnio replied:
Updated :)

LindvedKrvang commented on src/model/entities/part.ts:
```diff
@@ -204,30 +204,28 @@ export class Part {
+      this.stopOverlappingPiecesOnSameLayer(unPlannedPiece)
     } else {
-      this.removePieceOnLayer(unPlannedPiece.layer)
+      this.removeOverlappingPiecesOnSameLayer(unPlannedPiece)
```
I'm quite sure this is going to have unintended side-effects i.e. wrong Piece behaviour. What does this have to do with the synchronize ingest flow?

KvelaGorrrrnio replied:
I don't see any current use cases where this will have any problems. But it is worth having in mind going forward :+1:

LindvedKrvang commented on src/model/entities/rundown.ts:
```diff
@@ -693,6 +698,9 @@ export class Rundown extends BasicRundown {
       throw new NotFoundException(`Unable to find segment for part with id '${partId}' in rundown ${this.id}.`)
     }
     const removedPart: Part | undefined = segment.removePart(partId)
+    if (removedPart?.isOnAir()) {
+      this.activeCursor = this.createCursor(this.activeCursor, { part: removedPart })
+    }
```
Is this to make the Unsynced copi of the OnAir Part be the one the Cursor is referencing? I.e. the update you did to `Segment.removePart()`?

KvelaGorrrrnio replied:
It is :)

LindvedKrvang reviewed (changes requested):
The request changes is only for making sure this PR isn't merged until we are certain that your `removeOverlappingPiecesOnSameLayer()` doesn't create unintended sideeffects

KvelaGorrrrnio committed:
refactor: Combined deleted parts and segments info arguments for IngestDataChangeService.emitEventsFromRundownSynchronizeResult.

KvelaGorrrrnio committed:
refactor: Renames Part.doesPiecesOverlap to .doPiecesOverlap.

KvelaGorrrrnio committed:
refactor: Uses DeletedInfo interface instead of type literal.

KvelaGorrrrnio committed:
refactor: Made internal methods on IngestRundownSynchronizer private.

KvelaGorrrrnio committed:
Merge branch 'staging' into SOF-2298/synchronize-only-ingest-flow

RasmusAlbrektsen reviewed (approved):
LGTM
